processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -889,15 +900,34 @@ public Metadata getMetadata()
       {
         final List<DimensionDesc> dimensions = getDimensions();
 
-        Map<TimeAndDims, Integer> facts = null;
-        if (descending && sortFacts) {
-          facts = ((ConcurrentNavigableMap<TimeAndDims, Integer>) getFacts()).descendingMap();
+        final Iterable<Map.Entry<TimeAndDims, Integer>> facts;
+        if (sorted && sortFacts) {
+          facts = descending ? ((ConcurrentNavigableMap<TimeAndDims, Integer>) getFacts()).descendingMap().entrySet()
+                             : getFacts().entrySet();
+        } else if (sorted) {
+          // Materialize and sort
+          final Comparator<TimeAndDims> comparator = descending
+                                                     ? Ordering.from(dimsComparator()).reverse()
+                                                     : dimsComparator();
+          List<Map.Entry<TimeAndDims, Integer>> factsList = Lists.newArrayList(getFacts().entrySet());
```

Comment:
this may be ok for on-heap index but will unexpectedly put everything on-heap even for any off-heap implementation of IncrementalIndex (for example https://github.com/druid-io/druid/pull/2847 )
one option would be to update QueryToolChest to not exercise this path when off-heap merging was requested.